keps/sig-node/4191-split-image-filesystem/README.md

```diff
@@ -0,0 +1,891 @@
+    + // Information of container filesystem(s).
+    + // This is an optional field if container and image
+    + // storage are separated.
+    + // Default will be to return this as empty.
+    + repeated FilesystemUsage container_filesystems = 2;
+}
+```
+
+It is expected of the CRI implementation to return a unique identifier for images and containers so the Kubelet can ask CRI if the objects are stored on separate disks.
+In the dedicated disk for container runtime, images_filesystem and container_fileystem will be set to the same value.
+
+The CRI implementation can set this as needed.  The image and container filesystems are both arrays so this provides some extensibility in case these are stored on multiple disks.  
+
+Container runtimes will need to implement ImageFsInfo
+
+- [CRI-O Implementation](https://github.com/cri-o/cri-o/blob/main/server/image_fs_info.go)
+- [Containerd implementation](https://github.com/containerd/containerd/blob/main/pkg/cri/server/imagefs_info.go)
+
+An Alpha-Beta goal would be to have an implementation of `crictl imagefsinfo` that can allow for more detailed reports of the image fs info.
+
+See [PR](https://github.com/kannon92/cri-tools/pull/1) for an example.
+
+### Stats Summary
+
+Stats Summary has a field called runtime and we will add a containerFs to the runtime field.
+
+```golang
+// NodeStats holds node-level unprocessed sample stats.
+type NodeStats struct {
+  ...
+// Reference to the measured Node.
+// Stats about the underlying container runtime.
+// +optional
+Runtime *RuntimeStats `json:"runtime,omitempty"`
+}
+
+// RuntimeStats are stats pertaining to the underlying container runtime.
+type RuntimeStats struct {
+// Stats about the underlying filesystem where container images are stored.
+// This filesystem could be the same as the primary (root) filesystem.
+// Usage here refers to the total number of bytes occupied by images on the filesystem.
+// +optional
+ImageFs *FsStats `json:"imageFs,omitempty"`
++ // Stats about the underlying filesystem where container's writeable layer is stored.
++ // This filesystem could be the same as the primary (root) filesystem or the ImageFS.
++ // Usage here refers to the total number of bytes occupied by the writeable layer on the filesystem.
++ // +optional
++ ContainerFs *FsStats `json:"containerFs,omitempty"`
+}
+```
+
+In this KEP, ContainerFs can either be the same as ImageFs or NodeFs.
+
+We will add a more detailed function for ImageFsStats in the Provider Interface
+
+```golang
+type containerStatsProvider interface {
+...
+ImageFsStats(ctx context.Context) (*statsapi.FsStats, *statsapi.FsStats, error)
+}
+```
+
+If we have a single image filesystem then ImageFs includes both writable and read-only layer.  In this case, `ImageFsStats` will return an identical object for ImageFs and ContainerFs.
+
+In a case where the container runtime does not return a container filesystem, we will assume that the image_filesystem=container_filesystem.
```

Comment:
what happens when runtime just returns containerfs and not imagefs? is that an error condition?

Reply:
This is more of an implementation detail for CRIO.  

So if `graphroot` is specified than containerfs/imagefs are on same disk.  

Storage throws an error if one tries to set `imagestore` without setting `graphroot`.  If one only specifies graphroot we assume that containerfs=imagefs.  

This probably should be an error condition but the implementation I have in mind is that I will always send both containerfs and imagefs.

Reply:
CRI-O prevents this but we could revisit this if other implementations allow this.